File: python/wordslines.py
#!/usr/bin/env python3

import sys
from collections import defaultdict

class Result:
    def __init__(self, words_at_line, word_count):
        self.words_at_line = words_at_line
        self.word_count = word_count


def VanillaSolution(infname):
    with open(infname, 'r') as f:
        lines = [line.strip() for line in f]

    nlines = len(lines)
    words_at_line = [0] * nlines
    word_count = defaultdict(int)

    for idx, line in enumerate(lines):
        line = str(line)
        words = line.split()
        words_at_line[idx] = len(words)
        for word in words:
            word = word.lower()
            word_count[word] += 1

    return Result(words_at_line, word_count)


def WriteResult(result, out1fname, out2fname):
    with open(out1fname, 'w') as f:
        for idx, wordcount in enumerate(result.words_at_line):
            line = f"{idx+1} {wordcount}\n"
            f.write(line)

    with open(out2fname, 'w') as f:
        for word, count, in result.word_count.items():
            line = f"{word} {count}\n"
            f.write(line)


def main():
    if len(sys.argv) < 4:
        print(f"usage: {sys.argv[0]} <input-filename> <output-filename-1> <output-filename-2>")
        return

    infname, out1fname, out2fname = sys.argv[1], sys.argv[2], sys.argv[3]

    result = VanillaSolution(infname)
    WriteResult(result, out1fname, out2fname)

if __name__ == "__main__":
    main()

File: cpp/wordslines.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <iterator>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

typedef int32_t LineCounter;
typedef vector<LineCounter> WordsAtLine;
typedef int32_t WordCounter;
typedef unordered_map<string, WordCounter> WordCounterMap;


struct Result {
  // words_at_line_[i] = number of words in line i.
  WordsAtLine words_at_line_;

  // Map: word -> count.
  WordCounterMap word_count_;
};


Result
VanillaSolution(ifstream & ifs) {
  Result result;
  vector<string> words;
  istringstream iss;
  for (string line; getline(ifs, line); ) {
    iss.clear();
    iss.str(line);
    words.assign(istream_iterator<string>{iss}, istream_iterator<string>{});
    result.words_at_line_.push_back(static_cast<LineCounter>(words.size()));
    for (auto & word : words) {
      transform(begin(word), end(word), begin(word), ::tolower);
      result.word_count_[word] += 1;
    }
  }
  return result;
}


void
WriteResult(Result const& result, ostream & os1, ostream & os2) {
  for (size_t i = 0; i < result.words_at_line_.size(); ++i) {
    os1 << i+1 << ' ' << static_cast<int>(result.words_at_line_[i]) << '\n';
  }
  for (auto const& kv : result.word_count_) {
    os2 << kv.first << ' ' << kv.second << '\n';
  }
}


int main(int argc, char * argv[]) {
  if (argc < 4) {
    cerr << "Usage: " << argv[0] << " <input-filename> <output-filename-1> <output-filename-2>\n";
    return 0;
  }
  char const * const filename = argv[1];
  ifstream ifs{filename, ios::binary};
  if (!ifs.is_open()) {
    cerr << "Failed to open file " << filename << '\n';
    return 0;
  }

  char const * const output_filename_1 = argv[2];
  char const * const output_filename_2 = argv[3];
  
  Result const result_vanilla = VanillaSolution(ifs);
  ofstream ofs1{output_filename_1};
  ofstream ofs2{output_filename_2};
  WriteResult(result_vanilla, ofs1, ofs2);
}
